### PR TITLE
Deprecate CESM ponds (tr_pond_cesm)

### PR DESCRIPTION
**Pull request by dabail10**
For detailed information about submitting Pull Requests (PRs) to the CICE-Consortium,
please refer to: <https://github.com/CICE-Consortium/About-Us/wiki/Resource-Index#information-for-developers>

## PR checklist
- [X] Short (1 sentence) summary of your PR: 
Deprecate the tr_ponds_cesm option.
- [X] Developer(s): 
dabail10 (D. Bailey)
- [X] Suggest PR reviewers from list in the column to the right.
- [X] Please copy the PR test results link or provide a summary of testing completed below.
Did a full test suite compared against the weekly baselines.
5945 measured results of 5945 total results
5856 of 5945 tests PASSED
0 of 5945 tests PENDING
0 of 5945 tests MISSING data
89 of 5945 tests FAILED

Of the 89 tests, the majority of the changed answers were alt05 related which had tr_pond_cesm = .true.

The remaining test failures were as follows. I'm not sure why.

FAIL cheyenne_intel_smoke_gx3_8x1_cmplogrest_dynpicard_reprosum_run10day_thread bfbcomp cheyenne_intel_smoke_gx3_6x4_dynpicard_reprosum_run10day different-data
FAIL cheyenne_pgi_smoke_gx3_8x1_cmplogrest_dynpicard_reprosum_run10day_thread bfbcomp cheyenne_pgi_smoke_gx3_6x4_dynpicard_reprosum_run10day different-data
FAIL cheyenne_pgi_smoke_gx1_144x2_gx1prod_long_run10year run -1 -1 -1
FAIL cheyenne_pgi_smoke_gx1_144x2_gx1prod_long_run10year test 
FAIL cheyenne_gnu_smoke_gx3_8x1_cmplogrest_dynpicard_reprosum_run10day_thread bfbcomp cheyenne_gnu_smoke_gx3_6x4_dynpicard_reprosum_run10day different-data

- How much do the PR code changes differ from the unmodified code? 
    - [ ] bit for bit
    - [ ] different at roundoff level
    - [X] more substantial 
Removes all tests with tr_ponds_cesm = .true.
- Does this PR create or have dependencies on Icepack or any other models?
    - [X] Yes
    - [ ] No
- Does this PR add any new test cases?
    - [ ] Yes
    - [X] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/. A test build of the technical docs will be performed as part of the PR testing.)
    - [X] Yes
    - [ ] No, does the documentation need to be updated at a later time?
        - [ ] Yes
        - [ ] No 
- [X] Please provide any additional information or relevant details below:
There is an option UNDEPRECATE_CESMPONDS to reverse this. It also requires adding tr_pond_cesm and restart_pond_cesm back to ice_in.

### CODE REVIEW

File: cicecore/cicedynB/dynamics/ice_transport_driver.F90 [hunk 1]
```diff
@@ -1602,8 +1612,13 @@ subroutine state_to_work (nx_block, ny_block,        &
                                         * trcrn(i,j,it     ,n)
                enddo
                enddo
+#ifdef UNDEPRECATE_CESMPONDS
             elseif (trcr_depend(it) == 2+nt_apnd .and. &
                     tr_pond_cesm .or. tr_pond_topo) then
+#else
```
Comment (eclare108213) [inline cicecore/cicedynB/dynamics/ice_transport_driver.F90:1618]: This is interesting.  I wonder whether the .and. or the .or. took precedence in the logic for this elseif statement.
i.e.
`(trcr_depend(it)==2+nt_apnd .and. tr_pond_cesm) .or. tr_pond_topo`
or
`trcr_depend(it)==2+nt_apnd .and. (tr_pond_cesm .or. tr_pond_topo)`
?
The latter is correct, but what was the code actually doing?

**Review by eclare108213 — APPROVED** [review]
Looks good to me.

File: cicecore/cicedynB/dynamics/ice_transport_driver.F90 (hunk 1 above)
Comment (phil-blain) [reply cicecore/cicedynB/dynamics/ice_transport_driver.F90:1618]: According to https://www.intel.com/content/www/us/en/develop/documentation/fortran-compiler-oneapi-dev-guide-and-reference/top/language-reference/expressions-and-assignment-statements/expressions/summary-of-operator-precedence.html the `.and.` would have precedence on the `.or.` so the code would do the first thing you listed:
```fortran
(trcr_depend(it)==2+nt_apnd .and. tr_pond_cesm) .or. tr_pond_topo
```
i.e. the wrong thing if I understand correctly.

**Comment by eclare108213** [comment]
I approved this PR, but looking a little more closely at the test results noted above, I think it would be good to understand why the failures are happening, at least for the dynpicard / reprosum option.  These changes shouldn't affect those tests, should they?

**Comment by phil-blain** [comment]
Here are the failures quoted above, one per line (easier to read):

```
FAIL cheyenne_intel_smoke_gx3_8x1_cmplogrest_dynpicard_reprosum_run10day_thread bfbcomp cheyenne_intel_smoke_gx3_6x4_dynpicard_reprosum_run10day different-data
FAIL cheyenne_pgi_smoke_gx3_8x1_cmplogrest_dynpicard_reprosum_run10day_thread bfbcomp cheyenne_pgi_smoke_gx3_6x4_dynpicard_reprosum_run10day different-data
FAIL cheyenne_pgi_smoke_gx1_144x2_gx1prod_long_run10year run -1 -1 -1
FAIL cheyenne_pgi_smoke_gx1_144x2_gx1prod_long_run10year test
FAIL cheyenne_gnu_smoke_gx3_8x1_cmplogrest_dynpicard_reprosum_run10day_thread bfbcomp cheyenne_gnu_smoke_gx3_6x4_dynpicard_reprosum_run10day different-data
```

In general I do not currently expect the VP solver as it stands on `main` to be BFB between different decompositions, even with `reprosum`. So these 3  failures are all expected:
```
FAIL cheyenne_intel_smoke_gx3_8x1_cmplogrest_dynpicard_reprosum_run10day_thread bfbcomp cheyenne_intel_smoke_gx3_6x4_dynpicard_reprosum_run10day different-data
FAIL cheyenne_pgi_smoke_gx3_8x1_cmplogrest_dynpicard_reprosum_run10day_thread bfbcomp cheyenne_pgi_smoke_gx3_6x4_dynpicard_reprosum_run10day different-data
FAIL cheyenne_gnu_smoke_gx3_8x1_cmplogrest_dynpicard_reprosum_run10day_thread bfbcomp cheyenne_gnu_smoke_gx3_6x4_dynpicard_reprosum_run10day different-data
```

I do not expect the code in this PR to change anything in this result, I would expect these tests to also fail without these changes.

My upcoming PR which I mentioned last week at the telecon should fix those.

File: cicecore/cicedynB/dynamics/ice_transport_driver.F90 (hunk 1 above)
Comment (eclare108213) [reply cicecore/cicedynB/dynamics/ice_transport_driver.F90:1618]: Thanks.  This line is in the state_to_work routine, which is only used for upwind advection.  Tracers may not work properly with upwind advection -- see https://github.com/CICE-Consortium/CICE/issues/267 and https://github.com/CICE-Consortium/CICE/issues/542. This faulty logic could be part of the reason why, but it probably doesn't fix the entire issue, which is that many tracers are not implemented at all for upwind.

**Comment by eclare108213** [comment]
Thanks @phil-blain!  So the outstanding tests are gx1prod, which might just be timing out, @dabail10?

**Comment by dabail10** [comment]
Good catch! I will fix this.

**Comment by dabail10** [comment]
The failure is more cryptic. Seems to be dying in icepack_mechred.F90. Trying to resubmit this test.

**Comment by dabail10** [comment]
Ok. Are we happy with the results here? Should we merge this before the other two?

**Comment by eclare108213** [comment]
I think we can merge and it's fine with me to do the CESM ponds deprecation PR(s) first, then the others.  Does it make sense to do all of them in Icepack first, then update the Icepack version here and merge this PR, then pull the new Icepack into the other PR? OR does it make more sense to merge the Icepack CESM ponds PR first, then update Icepack here and merge, then merge the Icepack 0-layer and krdg PR, update Icepack again and finally merge the CICE 0-layer and krdg PR?

**Comment by apcraig** [comment]
If everything is just about ready, we could merge the two PRs together first in Icepack, then in CICE (after updating the Icepack submodule in one of the CICE PRs).  But, it might be better to keep the two distinct set of changes separate in case they don't play well together, in case we want to undo one of them, or if there might be a few days between readiness.  So, if CESM ponds is ready, we should merge the icepack PR, update the CICE submodule in the CICE PR and merge the CICE PR.  Then we would do the same with the 0layer/ridging PRs.

Is there any additional testing required for the CESM ponds PRs?  Would you like me to test independently?  Happy not to if others feel it's ready.  I'll do a quick review now.

**Comment by apcraig** [comment]
Looks like testing is fine.  alt05 will change answers and the dynpicard is failing in the baseline.  It's always good to look at the baseline results to check whether the failures are new.  Have you tested with UNDEPRECATE_CESMPONDS?  Not saying it's required, just curious.

Is there any other documentation of CESM ponds that needs to be removed from the user guide?

**Review by apcraig — APPROVED** [review]

**Pull request merged**